API for `LightClientBootstrap`, `LightClientFinalityUpdate`, `LightClientOptimisticUpdate` and light client events

GeemoCandama opened the pull request:
## Issue Addressed

#4894 and partially #3651 

## Proposed Changes

This PR implements the following Beacon APIs:
- new: [`/eth/v1/beacon/light_client/bootstrap/{block_root}`](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getLightClientBootstrap)
- new: [`/eth/v1/beacon/light_client/optimistic_update`](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getLightClientOptimisticUpdate)
- new: [`/eth/v1/beacon/light_client/finality_update`](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Beacon/getLightClientFinalityUpdate)
- updated: Add `light_client_optimistic_update` and `light_client_finality_update`  topics to [`/eth1/v1/events`](https://ethereum.github.io/beacon-APIs/?urls.primaryName=dev#/Events/eventstream).

## Additional Info

When the `--light_client_server` is not enabled a generic `404 Not Found` is returned, whereas when the server is enabled but the server has no latest update stored, a `404` with the message saying that No `LightClientOptimisticUpdate` is available.

GeemoCandama committed:
rebase and add comment

GeemoCandama committed:
conditional test

GeemoCandama committed:
test

GeemoCandama committed:
optimistic chould be working now

GeemoCandama committed:
finality should be working now

GeemoCandama committed:
try again

GeemoCandama committed:
try again

GeemoCandama committed:
clippy fix

GeemoCandama committed:
add lc bootstrap beacon api

GeemoCandama committed:
add lc optimistic/finality update to events

GeemoCandama committed:
fmt

GeemoCandama committed:
That error isn't occuring on my computer but I think this should fix it

dapplion commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1726,6 +1740,169 @@ pub fn serve<T: BeaconChainTypes>(
+                            warp_utils::reject::custom_server_error(
+                                "Failed to create light client bootstrap".to_string(),
+                            )
+                        })?;
```
Note: as expensive as ReqResp bootstrap handler. However both are disabled by default so do not add risk to existing use-cases 

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/network/src/network_beacon_processor/rpc_methods.rs#L265

jimmygchen replied:
Yep! This route is also disabled by default.

dapplion commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -1726,6 +1740,169 @@ pub fn serve<T: BeaconChainTypes>(
+            |chain: Arc<BeaconChain<T>>, accept_header: Option<api_types::Accept>| {
+                blocking_task(move || {
+                    let update = chain
+                        .latest_seen_optimistic_update
+                        .lock()
+                        .clone()
+                        .ok_or_else(|| {
+                            warp_utils::reject::custom_not_found(
+                                "No LightClientOptimisticUpdate is available".to_string(),
+                            )
+                        })?;
```
This strategy works in most cases but depends on the gossip handler to populate the latest_seen_optimistic_update variable. 

https://github.com/sigp/lighthouse/blob/441fc1691b69f9edc4bbdc6665f3efab16265c9b/beacon_node/beacon_chain/src/light_client_optimistic_update_verification.rs#L131

A future PR should eagerly compute the updates on each block.

dapplion commented on common/eth2/src/lib.rs:
```diff
@@ -562,6 +562,40 @@ impl BeaconNodeHttpClient {
+    /// Returns `Ok(None)` on a 404 error.
+    pub async fn get_beacon_light_client_optimistic_update<T: EthSpec>(
+        &self,
+    ) -> Result<Option<LightClientOptimisticUpdate<T>>, Error> {
```
What are the typing guarantees with this route? Is this library being used publicly? If someone attempts to use this client against Lodestar or Nimbus, it will error since the SSZ types are not correct for any fork (including capella)

jimmygchen replied:
Yeah this is pretty outdated, I've got some fixes for these on [my other branch](https://github.com/jimmygchen/lighthouse/pull/13/files#diff-0f211f7b56d5f17309287521594635a7863467e1d706046dd1ee79a06437bcc2R2189-R2201), will cherry-pick some over.

jimmygchen replied:
Fixed in https://github.com/sigp/lighthouse/pull/3954/commits/ad3cc69205ea312f0b87c683650e40a8135e8b5b

dapplion reviewed (commented):
The routes added by this PR follow the spec at a time. It is not compliant with current spec as the types have not been updated to use the new lightclient header type.

Since all server routes are behind a flag this PR is technically a strict improvement over unstable. The types can be fixed in a follow-up PR. Left a note on the eth2 lib, which will expose a broken client available by default.

I would suggest to fix conflicts and merge as-is, and optionally delete the http client routes

jimmygchen committed:
Merge branch 'unstable' into light_client_beacon_api_1

# Conflicts:
#	beacon_node/beacon_chain/src/events.rs
#	beacon_node/http_api/src/lib.rs
#	beacon_node/http_api/src/test_utils.rs
#	beacon_node/http_api/tests/main.rs
#	beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
#	beacon_node/lighthouse_network/src/rpc/methods.rs
#	beacon_node/lighthouse_network/src/service/api_types.rs
#	beacon_node/network/src/beacon_processor/worker/rpc_methods.rs
#	beacon_node/tests/test.rs
#	common/eth2/src/types.rs
#	lighthouse/src/main.rs

jimmygchen committed:
Add missing test file

jimmygchen committed:
Update light client types to comply with Altair light client spec.

jimmygchen committed:
Fix test compilation

jimmygchen committed:
Merge branch 'unstable' into light_client_beacon_api_1

dapplion commented on consensus/types/src/light_client_bootstrap.rs:
```diff
@@ -37,13 +41,31 @@ impl<T: EthSpec> LightClientBootstrap<T> {
+        match fork_name {
+            ForkName::Altair => Ok(serde_json::from_value::<LightClientBootstrap<T>>(value)
+                .map_err(serde::de::Error::custom))?,
+            ForkName::Base | ForkName::Merge | ForkName::Capella | ForkName::Deneb => {
```
bellatrix (Merge) variant is equal to altair variant so should be supported. The header was extended on capella

https://github.com/ethereum/consensus-specs/blob/dev/specs/capella/light-client/fork.md#upgrading-light-client-data

jimmygchen replied:
Nice catch! Fixed in https://github.com/sigp/lighthouse/pull/3954/commits/d23bba7ce06902e96764290a7eb6c0464362bd44

dapplion commented on beacon_node/http_api/src/lib.rs:
```diff
@@ -2385,6 +2400,177 @@ pub fn serve<T: BeaconChainTypes>(
+             block_root: Hash256,
+             accept_header: Option<api_types::Accept>| {
+                task_spawner.blocking_response_task(Priority::P1, move || {
+                    let state_root = chain
+                        .get_blinded_block(&block_root)
+                        .map_err(|_| {
+                            warp_utils::reject::custom_server_error(
+                                "Error retrieving block".to_string(),
+                            )
+                        })?
+                        .map(|signed_block| signed_block.state_root())
+                        .ok_or_else(|| {
+                            warp_utils::reject::custom_not_found(
+                                "Light client bootstrap unavailable".to_string(),
+                            )
+                        })?;
+
+                    let mut state = chain
+                        .get_state(&state_root, None)
+                        .map_err(|_| {
+                            warp_utils::reject::custom_server_error(
+                                "Error retrieving state".to_string(),
+                            )
+                        })?
+                        .ok_or_else(|| {
+                            warp_utils::reject::custom_not_found(
+                                "Light client bootstrap unavailable".to_string(),
+                            )
+                        })?;
+                    let fork_name = state
+                        .fork_name(&chain.spec)
+                        .map_err(inconsistent_fork_rejection)?;
+                    let bootstrap =
+                        LightClientBootstrap::from_beacon_state(&mut state).map_err(|_| {
+                            warp_utils::reject::custom_server_error(
+                                "Failed to create light client bootstrap".to_string(),
+                            )
+                        })?;
```
nit: returning a bootstrap upon request logic is necessary for both rest api and reqresp p2p. Should this code be moved to the chain to reduce a future diff?

jimmygchen replied:
Yep, good point - we also have the same logic in p2p below, I'll move it to the `BeaconChain`.
https://github.com/sigp/lighthouse/blob/4ce01ddd11b70bf6bdd6ac3896e2b33acbf00f72/beacon_node/network/src/network_beacon_processor/rpc_methods.rs#L300

jimmygchen replied:
Fixed in https://github.com/sigp/lighthouse/pull/3954/commits/d90df3f103fe68e9b49d7a81fd04a70e4b3519c1. This actually revealed a few more issues - we weren't returning a `ForkVersionedResponse` (wrapper with the `version` and `data` fields) in the API (fixed) and also not adding context bytes to the p2p responses (fixed in #3849).

dapplion reviewed (approved):
Impl is correct and good except two nits!

jimmygchen committed:
Support deserializing light client structures for the Bellatrix fork

jimmygchen committed:
Move `get_light_client_bootstrap` logic to `BeaconChain`. `LightClientBootstrap` API to return `ForkVersionedResponse`.

dapplion reviewed (approved):
Everything looks good now, excellent! 

Surfacing the two CI errors for others, both appear to be unrelated?

_[compile-with-beta-compiler](https://github.com/sigp/lighthouse/actions/runs/6899736366/job/18771805388#logs)_
```
 Compiling slot_clock v0.2.0 (/home/runner/work/lighthouse/lighthouse/common/slot_clock)
error: unused import: `std::time::SystemTimeError`
 --> common/slot_clock/src/system_time_slot_clock.rs:5:9
  |
5 | pub use std::time::SystemTimeError;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `-D unused-imports` implied by `-D warnings`
  = help: to override `-D warnings` add `#[allow(unused_imports)]`

error: could not compile `slot_clock` (lib) due to previous error
```

_[check-code](https://github.com/sigp/lighthouse/actions/runs/6899736366/job/18771804765#logs)_
```
error: unnecessary map_err on constructor Err(_)
   --> beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs:380:14
    |
380 |         _ => Err(err).map_err(RPCError::from),
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Err(RPCError::from(err))`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_map_on_constructor
    = note: `-D clippy::unnecessary-map-on-constructor` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::unnecessary_map_on_constructor)]`

   Compiling ethers-contract-derive v1.0.2
error: could not compile `lighthouse_network` (lib) due to previous error
```

jimmygchen commented:
Yeah these are related to the new Rust release (fix in #4931) and new lint warnings from the new beta compiler version (fix in #4932)

jimmygchen committed:
Misc fixes.
- log cleanup
- move http_api config mutation to `config::get_config` for consistency
- fix light client API responses

jimmygchen committed:
Add light client bootstrap API test and fix existing ones.

jimmygchen committed:
Merge branch 'unstable' into light_client_beacon_api_1

jimmygchen committed:
Fix test for `light-client-server` http api config.

jimmygchen committed:
Appease clippy

jimmygchen reviewed (approved):
I've made a couple of changes to bring this PR up to date with `unstable`. I think it looks good to me now - would be good to have this merged first, as there are a few parallel work streams on light client right now (caching optimisation, capella types etc) that may depend on some of the fixes here.

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -6446,6 +6446,39 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            return Ok(None);
+        };
+
+        let Some(mut state) = self.get_state(&state_root, None)? else {
```
It would be marginally more efficient to provide the block's slot as a hint here: `Some(block.slot())`.

jimmygchen replied:
Nice! good to know. I've updated here https://github.com/sigp/lighthouse/pull/3954/commits/9e1e126002253f825a033cd0b5b53700c703ffab

michaelsproul reviewed (approved):
Looks great! I had one very minor perf suggestion, but sounds like we'll be re-working and optimising this anyway so happy to merge without it!

jimmygchen committed:
Efficiency improvement when retrieving beacon state.

michaelsproul commented:
Rad, lets merge once CI passes.
